drivers/clock_control/clock_stm32_ll_h7.c

```diff
@@ -231,47 +331,121 @@ static int stm32_clock_control_init(struct device *dev)
 	while (LL_RCC_HSE_IsReady() != 1) {
 	}
 
-	/* Set FLASH latency */
-	LL_FLASH_SetLatency(LL_FLASH_LATENCY_4);
-
 	/* Main PLL configuration and activation */
 	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSE);
+
+	pllsrc_clock = HSE_VALUE;
+
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
+	/* Support for CSI oscillator */
+
+	LL_RCC_CSI_Enable();
+	while (LL_RCC_CSI_IsReady() != 1) {
+	}
+
+	/* Main PLL configuration and activation */
+	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_CSI);
+
+	pllsrc_clock = CSI_VALUE;
+
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI)
+	/* By default choose HSI as PLL clock source */
+
+	/* @TODO for USB support developer needs to force PLL must be clock from HSE */
+
+	/* Enable HSI oscillator */
+	LL_RCC_HSI_Enable();
+	while (LL_RCC_HSI_IsReady() != 1) {
+	}
+
+	/* Calibrate the HSI */
+	LL_RCC_HSI_SetCalibTrimming(32);
+	/* @TODO make HSI divider configurable */
+	LL_RCC_HSI_SetDivider(LL_RCC_HSI_DIV1);
+
+	/* Main PLL configuration and activation */
+	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSI);
+
+	pllsrc_clock = HSI_VALUE;
+
 #else
-	#error "CONFIG_CLOCK_STM32_PLL_SRC_HSE not selected"
-#endif /* CONFIG_CLOCK_STM32_PLL_SRC_HSE */
+
+	/* No clock source selected for PLL, by default, disable the PLL */
+	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_NONE);
```

Comment:
Shouldn't it be an error in this case ?

Reply:
That is a good question. I was wondering about adding a CONFIG_CLOCK_STM32_PLL_SRC_NONE option as there exists a LL_RCC_PLLSOURCE_NONE option. Having that option would mean the user would explicitely want the PLL to not be clocked by any frequency. My idea was that if no pllsrc clock is configured, the PLL wouldn't be clocked at all to save power.

Reply:
> That is a good question. I was wondering about adding a CONFIG_CLOCK_STM32_PLL_SRC_NONE option as there exists a LL_RCC_PLLSOURCE_NONE option. Having that option would mean the user would explicitly want the PLL to not be clocked by any frequency.

I'm afraid having a CONFIG_CLOCK_STM32_PLL_SRC_NONE would confuse users.
Besides I'm not sure how this could work as we need a clock to be configured and current configuration scheme is pretty exclusive (HSI or HSE or .. or PLL, and this is a compile time choice). So if you chose PLL, you don't have other working clock configured, and if PLL has no source, you just have no clock.
So I don't see a specific interest, specially with current clock configuration system.

> My idea was that if no pllsrc clock is configured, the PLL wouldn't be clocked at all to save power.

This should be the case already with existing configurations. If clock source is set to HSE/HSI/..., PLL should not be started (and vice-versa).

> But most peripheral like CAN, SPI123 and SAI peripherals are clocked from the PLL1Q which makes the PLL necessary for most of the applications.

Being high speed and precise, PLL is a quite useful clock. But some might have good reasons not to use it, so let's not make this kind of assumptions.

Reply:
Okey, let me try to sum it up. 

My implementation allow the user to:

- select a system clock source: HSI, HSE, CSI or PLL:
  - this should mean that all of the clocks that depend on SYSCLK will derived from the chosen oscillator
  - but chosing HSI/HSE or CSI for SYSCLK directly shouldn't imply that the PLL stays OFF, because CAN and SPI devices could depend on the PLL1Q to be running 
    - --> **that is something I'm not doing now and chosing HSI/HSE or CSI doesn't enable the PLL if CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL is not set**
    - --> **I suggest to enable the PLL if any PLLSRC clock is configured and only setting it as SYSCLK source if CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL is set.**
    - This means enabling the PLL just after it was configured, hence moving the PLL1_Enable code just after the PLL dividers configuration. So if any clock source is provided for the PLL, it will be activated.
    - STM32CubeMX doesn't really allow None clock source for the PLL. 
    - If a user doesn't want to use the PLL and keep it deactivated, then we need to make sure that the error case would be that he selected the PLL as SYSCLK source. This would happen if none of the CONFIG_CLOCK_PLL_SRC_HSI/HSE/CSI are set *AND* CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL is set. **This would lead to no system clock, this is an error.**
    - Having no clock at the PLL should be fine as long as we have HSE/HSI/CSI as SYSCLK_SRC.

So I think that c**onfiguring the PLL should be independent from configuring the SYSCLK** source. And that configuring the PLL shouldn't be impacted by the choice of SYSCLK, unless you choose to disable the PLL but use it as SYSCLK.

Tell me what you think about that 😄

Reply:
BTW I have just tried setting the SYSCLK to PLL and not specifying any PLL_SRC value. the Kconfig system defaults it to PLL_SRC_HSI and adds the dividers values to create a 129MHz clock

Reply:
@lochej I get your point about letting user configure whatever source clock for core and PLL for peripherals.
This is indeed not the current configuration scheme that has been used so far. 
I see two cons in that approach right now:
- As you say this is a deviation vs other STM32
- This is currently done in peripheral drivers (cf i2s code for  instance).

So I'm not fundamentally against this proposal but I prefer:
- it is done in a dedicated PR
- It takes into account current peripherals implementation (ie rework if needed)
- it propose same behavior for the whole STM32 family

Reply:
@erwango That is fair, we'll keep with the current scheme for this PR. If we stick to it, then the #else statement is not an error but just the case in which the PLL is not configured.

Reply:
And I'd say that this should actually be an error... but after check, same issue applies to other stm32 implementation, so I won't block

Reply:
> 
> 
> And I'd say that this should actually be an error... but after check, same issue applies to other stm32 implementation, so I won't block

I can't understand why this could be an error and how we could handle it, because in the end, the user just didn't select any PLLSRC so everything is fine as long as we don't use the PLL as a clock for any peripheral. And you are right, that is the way it is handled in other STM32 implementation.

Reply:
This is an issue because user selects `CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y` (with help message "Use PLL as source of SYSCLK") and this is not what code is doing.

Reply:
I'm really sorry but I still don't get it 🙃. The thing is that if the user selects ```CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y``` then a PLLCLK source must be configured. This is always the case by present Kconfig scheme which enables the ```CONFIG_CLOCK_STM32_PLL_SRC_HSI=y``` hence giving a PLLSRC clock by default. If the SYSCLK_SRC_PLL symbol is not configured, then none of the PLL config will be compiled other than that code which disables the PLLSRC clock.

My code has 2 sections, first I configure the PLL if configuration symbols for they are present, then I select the SYSCLK clock source. The error you indicate relates to the configuration of the PLL which in a sense doesn't correspond to the setting of the SYSCLK.

What this code actually does is disable the PLL if it is not selected. If that is a problem though, I can remove the #else statement and the setting of the PLL to None.

Reply:
@lochej, my bad I had forgot `CLOCK_STM32_SYSCLK_SRC_PLL` is achoice and comes with a default value and then having no PLL configuration is not an expected case. This is fine, sorry for the trouble.

Reply:
> @lochej, my bad I had forgot `CLOCK_STM32_SYSCLK_SRC_PLL` is achoice and comes with a default value and then having no PLL configuration is not an expected case. This is fine, sorry for the trouble.

Now we agree on this one 😄 Thanks for resolving.